Flexify the winning condition

paul90317 opened the pull request:
I introduced the macro `GOAL`, allowing players to win by achieving a line of `GOAL` consecutive marks instead of only 3.

Additionally, I introduced the `line_t` struct to reduce code duplication in the `get_score` and `check_win` functions.

paul90317 committed:
Flexify the winning condition

I introduced the macro `GOAL`, allowing players to win by
achieving a line of `GOAL` consecutive marks instead of only 3.

Additionally, I introduced the `line_t` struct to reduce code
duplication in the `get_score` and `check_win` functions.

jserv commented on ttt.c:
```diff
@@ -10,16 +10,26 @@
 
 #define MAX_SEARCH_DEPTH 6
 #define BOARD_SIZE 4
+#define GOAL_NUMBER 3
```
Simply rename it `GOAL`.

jserv commented on ttt.c:
```diff
@@ -10,16 +10,26 @@
-static int no_step[2] = {0, 0};
-static const int step_forward[2] = {1, 2}, step_backward[2] = {-1, -2};
+typedef struct {
+    int i_shift, j_shift, i_lower_bound, j_lower_bound, i_upper_bound,
```
Split into two statements of declarations.

jserv commented on ttt.c:
```diff
@@ -10,16 +10,26 @@
+    {1, 1, 0, 0, BOARD_SIZE - GOAL_NUMBER + 1,
+     BOARD_SIZE - GOAL_NUMBER + 1},  // PRIMARY
+    {1, -1, 0, GOAL_NUMBER - 1, BOARD_SIZE - GOAL_NUMBER + 1,
+     BOARD_SIZE}  // SECONDARY
```
Place a comma `,` right after `BOARD_SIZE}`, so that clang-format can indent better.

jserv commented on ttt.c:
```diff
@@ -112,71 +122,27 @@ void draw_board(const char *t)
-                t[GET_INDEX(i + i_shift[0], j + j_shift[0])] ==
-                    t[GET_INDEX(i + i_shift[1], j + j_shift[1])]) {
-                return t[GET_INDEX(i, j)];
+    for (int i_line = 0; i_line < 4; ++i_line) {
```
Can you avoid deeper nested loops?

visitorckw commented:
If the goal number is 3, but we have >= 4 consecutive pieces of the same color, should it be considered a victory? Or should the game continue, similar to the overline rule in games like Gomoku?

paul90317 commented:
> If the goal number is 3, but we have >= 4 consecutive pieces of the same color, should it be considered a victory? Or should the game continue, similar to the overline rule in games like Gomoku?

It will be considered a victory. I didn't know Gomoku had such a rule until you said that.

visitorckw commented:
> > If the goal number is 3, but we have >= 4 consecutive pieces of the same color, should it be considered a victory? Or should the game continue, similar to the overline rule in games like Gomoku?
> 
> It will be considered a victory. I didn't know Gomoku had such a rule until you said that.

Yeah, I know the implementation considers it a victory. I just wanted to take this opportunity to discuss which rule we should stick to. I don't have a strong opinion on either side.

paul90317 commented:
Me too, but I think we can introduce a macro `ALLOW_EXCEED` to let users decide which rule type want.

visitorckw commented:
> Me too, but I think we can introduce a macro `ALLOW_EXCEED` to let users decide which rule type want.

Sounds like a good idea to me :)

paul90317 commented:
Thanks :), however designing a  better heuristic function of AI for new rules is a challenge.

visitorckw commented:
> Thanks :), however designing a better heuristic function of AI for new rules is a challenge.

Mmm, that's something I haven't thought about. Maybe we can stick to the current rules for now?

visitorckw commented:
Additionally, I believe the Game Rules section of the README needs to be updated accordingly.

jserv commented:
> Me too, but I think we can introduce a macro `ALLOW_EXCEED` to let users decide which rule type want.

Add the corresponding `_Static_assert` for build-time checks.

paul90317 committed:
Decide if the player can win with exceeding marks

I introduced macro `ALLOW_EXCEED` to let programmers decide whether the
player can win with exceeding marks. This rule is inspired by Gomoku.

Link: https://en.wikipedia.org/wiki/Gomoku

paul90317 commented:
I have introduced `ALLOW_EXCEED` and added the corresponding `_Static_assert`.

jserv commented on ttt.c:
```diff
@@ -10,16 +10,35 @@
+         ? (else_value)                                             \
+         : (table)[GET_INDEX(i, j)])
+
+_Static_assert((BOARD_SIZE) <= 26, "Board size must not be greater than 26");
```
It is not necessary to have extra parentheses. That is, write as `_Static_assert(BOARD_SIZE <= 26, ...)`.

paul90317 committed:
Modify README.md and remove extra parentheses

In this commit, I rewrote the "Game Rules" part of README.md with
build-time game parameters `GOAL` and `ALLOW_EXCEED`.

I also removed the extra parentheses.

jserv commented on ttt.c:
```diff
@@ -10,16 +10,34 @@
+#define GOAL 3
+#define ALLOW_EXCEED 1
+#define N_GRIDS (BOARD_SIZE * BOARD_SIZE)
+#define GET_INDEX(i, j) ((i) *BOARD_SIZE + (j))
```
Add parentheses to `BOARD_SIZE`, so that clang-format can indent `GET_INDEX` better. Clang-format is a great tool, but it is sometimes not smart enough.

paul90317 replied:
Okay, I gradually realize...

jserv commented:
Thank @paul90317 for contributing. The git commit messages were amended.
